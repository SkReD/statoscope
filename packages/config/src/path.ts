import path from 'path';
import module from 'module';
import chalk from 'chalk';

export enum PackageAliasType {
  PLUGIN = 'PLUGIN',
  REPORTER = 'REPORTER',
}

export function normalizePath(source: string, rootDir: string): string {
  return source.replace('<rootDir>', rootDir);
}

export function makeRequireFromPath(pathname: string): NodeRequire {
  return module.createRequire(path.resolve(pathname, '_'));
}

// package namespace (empty string when not set) followed by the package name (wo namespace)
type PackageName = [string, string];

const packageAliasPrefixes = new Map<PackageAliasType, Record<PrefixType, string>>();

enum PrefixType {
  STATOSCOPE_NS = '@statoscope',
  NON_STATOSCOPE_NS = '@NON_STATOSCOPE_NS',
}

packageAliasPrefixes.set(PackageAliasType.PLUGIN, {
  [PrefixType.STATOSCOPE_NS]: 'stats-validator-plugin',
  [PrefixType.NON_STATOSCOPE_NS]: 'statoscope-stats-validator-plugin',
});

packageAliasPrefixes.set(PackageAliasType.REPORTER, {
  [PrefixType.STATOSCOPE_NS]: 'stats-validator-reporter',
  [PrefixType.NON_STATOSCOPE_NS]: 'statoscope-stats-validator-reporter',
});

/**
 * Concatenate namespace (when provided) with the package name.
 */
function getFullPackageName([namespace, name]: PackageName): string {
  if (namespace) {
    return `${namespace}/${name}`;
  }

  return name;
}

/**
 * Return error message based on the list of package aliases (provided by the user and alternative).
 */
function getAliasPackageResolutionError(packageAliases: PackageName[]): string {
  const providedAlias = getFullPackageName(packageAliases[0]);

  let errorMessage = `Can't resolve package ${chalk.yellow.italic(providedAlias)}.\n`;

  const { italic: italicChalk } = chalk;

  if (packageAliases.length > 1) {
    const alternativeAlias = getFullPackageName(packageAliases[1]);

    errorMessage += `Also tried to resolve it with ${italicChalk.yellow(
      alternativeAlias,
    )} alias which didn't work either.\n\n`;
  } else {
    errorMessage += '\n\n';
  }

  errorMessage += 'Try installing the package locally:\n';
  errorMessage += `- with ${italicChalk('npm')}: ${chalk.bgGrey(
    'npm i -D ' + providedAlias,
  )} (or corresponding package alias)\n`;
  errorMessage += `- with ${italicChalk('yarn')}: ${chalk.bgGrey(
    'yarn add -D ' + providedAlias,
  )} (or corresponding package alias)\n`;

  return errorMessage;
}

/**
 * Resolve full or short package name to it's absolute path.
 * @param packageAliasType - Plugin or reporter enum.
 * @param aliasName - Package name (alias). Can be passed in the "full" or "short" form.
 *   Will append prefixes based on packageAliasType value in order to resolve the short form alias.
 * @param fromDir - Directory used as "root" while resolving the package.
 */
export function resolveAliasPackage(
  packageAliasType: PackageAliasType,
  aliasName: string,
  fromDir: string,
): string {
  const localRequire = makeRequireFromPath(fromDir);
  aliasName = normalizePath(aliasName, fromDir);

  if (aliasName.startsWith('.') || path.isAbsolute(aliasName)) {
    localRequire(aliasName);
    return aliasName;
  }

  let packageNamespace = '';
  let packageName = aliasName;

  const packageNameRegex = /^(@.+?)\/(.+)/;

  if (packageNameRegex.test(aliasName)) {
    [, packageNamespace = '', packageName] = aliasName.match(packageNameRegex)!;
  }

  // can have from one up to two entries
  const packageAliases: PackageName[] = [
    [packageNamespace, packageName], // original form
  ];

  const prefixes = packageAliasPrefixes.get(packageAliasType)!;

  const prefix =
    packageNamespace === PrefixType.STATOSCOPE_NS
      ? prefixes[PrefixType.STATOSCOPE_NS]
      : prefixes[PrefixType.NON_STATOSCOPE_NS];

  if (!packageName.startsWith(prefix)) {
    packageAliases.push([packageNamespace, `${prefix}-${packageName}`]);
  }

  const paths = packageAliases.map(([namespace, packageName]) =>
    path.join(namespace, packageName),
  );

  for (const path of paths) {
    try {
      localRequire(path);
      return path;
      // eslint-disable-next-line no-empty
    } catch (e) {}
  }

  throw new Error(getAliasPackageResolutionError(packageAliases));
}
