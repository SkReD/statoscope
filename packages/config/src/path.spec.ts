import path from 'path';
import { normalizePath, PackageAliasType, resolveAliasPackage } from './path';

const rootDir = path.resolve(__filename, '../../../..');
const fixtureDir = path.resolve(__filename, '../../../../test/fixtures/config');

function normalize(path: string): string {
  return path.replace(rootDir, '');
}

const { PLUGIN: pluginType } = PackageAliasType;

describe('resolveAliasPackage', () => {
  describe('plugins behavior', () => {
    describe('resolves by full package name (no need to use prefixes)', () => {
      test('with random namespace and random name', () => {
        expect(
          normalize(resolveAliasPackage(pluginType, '@foo/bar', fixtureDir)),
        ).toMatchInlineSnapshot('"@foo/bar"');
      });

      test('with random namespace and expected (prefixed) name', () => {
        expect(
          normalize(
            resolveAliasPackage(
              pluginType,
              '@foo/statoscope-stats-validator-plugin-1',
              fixtureDir,
            ),
          ),
        ).toMatchInlineSnapshot('"@foo/statoscope-stats-validator-plugin-1"');
      });

      test('without namespace and random name', () => {
        expect(
          normalize(resolveAliasPackage(pluginType, 'foo', fixtureDir)),
        ).toMatchInlineSnapshot('"foo"');
      });

      test('without namespace and expected (prefixed) name', () => {
        expect(
          normalize(
            resolveAliasPackage(
              pluginType,
              'statoscope-stats-validator-plugin-1',
              fixtureDir,
            ),
          ),
        ).toMatchInlineSnapshot('"statoscope-stats-validator-plugin-1"');
      });

      test('with @statoscope namespace and random name', () => {
        expect(
          normalize(resolveAliasPackage(pluginType, '@statoscope/foo', fixtureDir)),
        ).toMatchInlineSnapshot('"@statoscope/foo"');
      });

      test('with @statoscope namespace and expected (prefixed) name', () => {
        expect(
          normalize(
            resolveAliasPackage(
              pluginType,
              '@statoscope/stats-validator-plugin-1',
              fixtureDir,
            ),
          ),
        ).toMatchInlineSnapshot('"@statoscope/stats-validator-plugin-1"');
      });
    });

    describe('resolves by short package name (which has to be prefixed)', () => {
      test('with @statoscope namespace and random name', () => {
        expect(
          normalize(resolveAliasPackage(pluginType, '@statoscope/1', fixtureDir)),
        ).toMatchInlineSnapshot('"@statoscope/stats-validator-plugin-1"');
      });

      test('with random namespace and random name', () => {
        expect(
          normalize(resolveAliasPackage(pluginType, '@foo/1', fixtureDir)),
        ).toMatchInlineSnapshot('"@foo/statoscope-stats-validator-plugin-1"');
      });

      test('wo namespace and with random name', () => {
        expect(
          normalize(resolveAliasPackage(pluginType, '1', fixtureDir)),
        ).toMatchInlineSnapshot('"statoscope-stats-validator-plugin-1"');
      });

      test("won't resolve without namespace and @statoscope specific prefix", () => {
        expect(() =>
          resolveAliasPackage(pluginType, 'wont-find', fixtureDir),
        ).toThrowErrorMatchingSnapshot();
      });

      test("won't resolve with @statoscope namespace and name with no (or custom) namespace prefix", () => {
        expect(() =>
          resolveAliasPackage(pluginType, '@statoscope/wont-find', fixtureDir),
        ).toThrowErrorMatchingSnapshot();
      });

      test("won't resolve with random namespace and @statoscope specific prefix", () => {
        expect(() =>
          resolveAliasPackage(pluginType, '@foo/wont-find', fixtureDir),
        ).toThrowErrorMatchingSnapshot();
      });
    });
  });

  test('throws when package is not found', () => {
    expect(() =>
      resolveAliasPackage(PackageAliasType.REPORTER, '@statoscope/2', fixtureDir),
    ).toThrowErrorMatchingSnapshot();
  });

  describe('relative path support', () => {
    test('resolves package by relative path', () => {
      expect(
        normalize(
          resolveAliasPackage(
            pluginType,
            './node_modules/@statoscope/stats-validator-reporter-1',
            fixtureDir,
          ),
        ),
      ).toMatchInlineSnapshot(`"./node_modules/@statoscope/stats-validator-reporter-1"`);
    });

    test('throws when package is not there', () => {
      expect(() =>
        resolveAliasPackage(
          pluginType,
          './node_modules/@statoscope/not-there',
          fixtureDir,
        ),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('absolute path support', () => {
    test('resolves package by absolute path', () => {
      expect(
        normalize(
          resolveAliasPackage(
            pluginType,
            path.join(
              fixtureDir,
              './node_modules/@statoscope/stats-validator-reporter-1',
            ),
            fixtureDir,
          ),
        ),
      ).toMatchInlineSnapshot(
        `"/test/fixtures/config/node_modules/@statoscope/stats-validator-reporter-1"`,
      );
    });

    test('throws when package is not there', () => {
      expect(() =>
        resolveAliasPackage(
          pluginType,
          path.join(fixtureDir, './node_modules/@statoscope/not-there'),
          fixtureDir,
        ),
      ).toThrow(); // can't use snapshot here because of the absolute path
    });
  });
});

describe('normalizePath', () => {
  test('basic functionality', () => {
    expect(normalizePath('<rootDir>/baz', 'foo/bar')).toBe('foo/bar/baz');
    expect(normalizePath('foo/bar/baz', 'fooo/barr')).toBe('foo/bar/baz');
  });
});
