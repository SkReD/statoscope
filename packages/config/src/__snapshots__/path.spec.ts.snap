// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveAliasPackage plugins behavior resolves by short package name (which has to be prefixed) won't resolve with @statoscope namespace and name with no (or custom) namespace prefix 1`] = `
"Can't resolve package @statoscope/wont-find.
Also tried to resolve it with @statoscope/stats-validator-plugin-wont-find alias which didn't work either.

Try installing the package locally:
- with npm: npm i -D @statoscope/wont-find (or corresponding package alias)
- with yarn: yarn add -D @statoscope/wont-find (or corresponding package alias)
"
`;

exports[`resolveAliasPackage plugins behavior resolves by short package name (which has to be prefixed) won't resolve with random namespace and @statoscope specific prefix 1`] = `
"Can't resolve package @foo/wont-find.
Also tried to resolve it with @foo/statoscope-stats-validator-plugin-wont-find alias which didn't work either.

Try installing the package locally:
- with npm: npm i -D @foo/wont-find (or corresponding package alias)
- with yarn: yarn add -D @foo/wont-find (or corresponding package alias)
"
`;

exports[`resolveAliasPackage plugins behavior resolves by short package name (which has to be prefixed) won't resolve without namespace and @statoscope specific prefix 1`] = `
"Can't resolve package wont-find.
Also tried to resolve it with statoscope-stats-validator-plugin-wont-find alias which didn't work either.

Try installing the package locally:
- with npm: npm i -D wont-find (or corresponding package alias)
- with yarn: yarn add -D wont-find (or corresponding package alias)
"
`;

exports[`resolveAliasPackage relative path support throws when package is not there 1`] = `"Cannot find module './node_modules/@statoscope/not-there' from '../../test/fixtures/config/_'"`;

exports[`resolveAliasPackage throws when package is not found 1`] = `
"Can't resolve package @statoscope/2.
Also tried to resolve it with @statoscope/stats-validator-reporter-2 alias which didn't work either.

Try installing the package locally:
- with npm: npm i -D @statoscope/2 (or corresponding package alias)
- with yarn: yarn add -D @statoscope/2 (or corresponding package alias)
"
`;
